src/components/CytoscapeGraph/Layout/BoxLayout.ts

```diff
@@ -0,0 +1,269 @@
+}
+
+/**
+ * Synthetic edge generator takes care of creating edges without repeating the same edge (targetA -> targetB) twice
+ */
+class SyntheticEdgeGenerator {
+  private nextId = 0;
+  private generatedMap = {};
+
+  public getEdge(source: any, target: any) {
+    const sourceId = this.normalizeToParent(source).id();
+    const targetId = this.normalizeToParent(target).id();
+
+    if (sourceId === targetId) {
+      return false;
+    }
+
+    const key = `${sourceId}->${targetId}`;
+
+    if (this.generatedMap[key]) {
+      return false;
+    }
+
+    this.generatedMap[key] = true;
+
+    return {
+      group: 'edges',
+      data: {
+        id: 'synthetic-edge-' + this.nextId++,
+        source: sourceId,
+        target: targetId
+      }
+    };
+  }
+
+  // Returns the parent if any or the element itself.
+  private normalizeToParent(element: any) {
+    return element.isChild() ? element.parent() : element;
+  }
+}
+
+/**
+ * Main class for the BoxLayout, used to bridge with cytoscape to make it easier to integrate with current code
+ */
+export default class BoxLayout {
+  readonly options;
+  readonly cy;
+  readonly elements;
+  readonly syntheticEdgeGenerator;
+
+  constructor(options: any) {
+    this.options = { ...options };
+    this.cy = this.options.cy;
+    this.elements = this.options.eles;
+    this.syntheticEdgeGenerator = new SyntheticEdgeGenerator();
+  }
+
+  /**
+   * This code gets executed on the cy.layout(...).run() is our entrypoint of this algorithm.
+   */
+  run() {
+    const { appBoxLayout, clusterBoxLayout, defaultLayout, namespaceBoxLayout } = this.options;
+    let allBoxNodes = this.cy.collection();
+    let removedElements = this.cy.collection();
+    let syntheticEdges = this.cy.collection();
+
+    for (const boxByType of [BoxByType.APP, BoxByType.NAMESPACE, BoxByType.CLUSTER]) {
+      const boxNodes = this.getBoxNodes(boxByType);
+      allBoxNodes = allBoxNodes.add(boxNodes);
+
+      let boxLayoutOptions = defaultLayout;
+      switch (boxByType) {
+        case BoxByType.APP:
+          if (appBoxLayout) {
+            boxLayoutOptions = appBoxLayout;
+          }
+          break;
+        case BoxByType.CLUSTER:
+          if (clusterBoxLayout) {
+            boxLayoutOptions = clusterBoxLayout;
+          }
+          break;
+        case BoxByType.NAMESPACE:
+          if (namespaceBoxLayout) {
+            boxLayoutOptions = namespaceBoxLayout;
+          }
+          break;
+      }
+
+      // (1.a) Prepare box node by assigning a size and running the compound layout
+      boxNodes.each(boxNode => {
+        const boxedNodes = boxNode.children();
+        const boxedElements = boxedNodes.add(boxedNodes.edgesTo(boxedNodes));
+        const boxLayout = boxedElements.layout(boxLayoutOptions);
+
+        // discrete layouts (dagre) stop before run() returns, continuous layouts (cose,cola) don't,
+        // force discrete behavior because we need each box to layout before continuing.
```

Comment:
> I think @josejulio comment still stands, because `runDiscrete` can return while the layout is still running.

You are correct there.

> Anyway, I'm not seeing any issue. I guess it's because it's always dagre inside the boxes, right?

If dagre is used in children, then it's fine, but I understood that you wanted to use the continuos layouts. If they work a raw guess is that It could also be that the continuous layouts do a first pass on their "run" method and that is enough for the purposes of this algorithm. Hard to tell without looking at their code :).

Reply:
In my testing, the current code waits as desired for the layout to stop, regardless of which layout is selected.  We force dagre only for app boxes, namespace and cluster boxing use the selected layout for the boxed nodes, as does the final layout for the full graph.  If you are not seeing this behavior let me know.

Reply:
@israel-hdez @josejulio OK, after talking to Edgar and doing some more reading I think I now have a better handle on what is needed to support a continuous layout for boxes.  I had been hoping for something more simple but have now accepted that it's hard.  :)  I have restructured the code and I'm fairly sure it now performs correctly.